File: 02_stack/02_evaluate_reverse_polish_notation/evaluate_reverse_polish_notation.cpp
#include <vector>
using std::vector;

#include <string>
using std::string;

#include <stack>
using std::stack;

#include <unordered_map>

class Solution
{
  public:
    int evalRPN(vector<string>& tokens)
    {
        stack<int> track;

        for (const string& token : tokens)
        {
            if (operations.count(token))
            {
                int b = track.top();
                track.pop();
                int a = track.top();
                track.pop();
                track.push(operations[token](a, b));
            }
            else
                track.push(std::stoi(token));
        }
        return track.top();
    }

  private:
    std::unordered_map<std::string, int (*)(int, int)> operations = {
        {"+", [](int a, int b) { return a + b; }},
        {"-", [](int a, int b) { return a - b; }},
        {"*", [](int a, int b) { return a * b; }},
        {"/", [](int a, int b) { return a / b; }}};
};

#include <iostream>

int main()
{
    vector<string> notations = {"2", "1", "+", "3", "*"};
    Solution       solution;

    int answer = solution.evalRPN(notations);
    std::cout << answer << "\n";
}

File: NeetCode/02_stack/02_evaluate_reverse_polish_notation/evaluate_reverse_polish_notation.py
from typing import List, Callable


class Solution:
    operations = {
        "+": lambda a, b: a + b,
        "-": lambda a, b: a - b,
        "*": lambda a, b: a * b,
        "/": lambda a, b: int(a / b)
    }

    def evalRPN(self, tokens: List[str]) -> int:
        track = list()
        for token in tokens:
            if token in "+-*/":
                second = track.pop()
                first = track.pop()
                track.append(self.operations[token](first, second))
            else:
                track.append(int(token))
        return track[0]


if __name__ == "__main__":
    notation = ["2", "1", "+", "3", "*"]
    notation = ["4", "13", "5", "/", "+"]
    notation = ["10", "6", "9", "3", "+", "-11",
                "*", "/", "*", "17", "+", "5", "+"]
    solution = Solution()
    print(solution.evalRPN(notation))
